cmd/kubeadm/app/phases/upgrade/versiongetter.go

```diff
@@ -0,0 +1,139 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	versionutil "k8s.io/kubernetes/pkg/util/version"
+	"k8s.io/kubernetes/pkg/version"
+)
+
+// versionGetter defines an interface for fetching different versions.
+// Easy to implement a fake variant of this interface for unit testing
+type versionGetter interface {
+	GetClusterVersion() (string, *versionutil.Version, error)
+	GetKubeadmVersion() (string, *versionutil.Version, error)
+	GetLatestKubeVersion(string, string) (string, *versionutil.Version, error)
+	GetKubeletVersions() (map[string]int32, error)
+}
+
+// kubeVersionGetter handles the version-fetching mechanism from external sources
+type kubeVersionGetter struct {
+	client clientset.Interface
+	w      io.Writer
+}
+
+// Make sure kubeVersionGetter implements the versionGetter interface
+var _ versionGetter = &kubeVersionGetter{}
+
+// NewKubeVersionGetter returns a new instance of kubeVersionGetter
+func NewKubeVersionGetter(client clientset.Interface, writer io.Writer) *kubeVersionGetter {
+	return &kubeVersionGetter{
+		client: client,
+		w:      writer,
+	}
+}
+
+// GetClusterVersion gets API server version
+func (g *kubeVersionGetter) GetClusterVersion() (string, *versionutil.Version, error) {
+	fmt.Fprintf(g.w, "[upgrade/versions] Cluster version: ")
+	clusterVersionInfo, err := g.client.Discovery().ServerVersion()
+	if err != nil {
+		fmt.Fprintln(g.w, "<notfound>")
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't fetch cluster version from the API Server: %v\n", err)
+		return "", nil, err
+	}
+	fmt.Fprintln(g.w, clusterVersionInfo.String())
+
+	clusterVersion, err := versionutil.ParseSemantic(clusterVersionInfo.String())
+	if err != nil {
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't parse cluster version: %v\n", err)
+		return "", nil, err
+	}
+	return clusterVersionInfo.String(), clusterVersion, nil
+}
+
+// GetKubeadmVersion gets kubeadm version
+func (g *kubeVersionGetter) GetKubeadmVersion() (string, *versionutil.Version, error) {
+	kubeadmVersionInfo := version.Get()
+	fmt.Fprintf(g.w, "[upgrade/versions] kubeadm version: %s\n", kubeadmVersionInfo.String())
+
+	kubeadmVersion, err := versionutil.ParseSemantic(kubeadmVersionInfo.String())
+	if err != nil {
+		fmt.Fprintf(g.w, "[upgrade/versions] FATAL: Couldn't parse kubeadm version: %v\n", err)
+		return "", nil, err
+	}
+	return kubeadmVersionInfo.String(), kubeadmVersion, nil
+}
+
+// GetKubeadmVersion gets the latest versions from CI
+func (g *kubeVersionGetter) GetLatestKubeVersion(ciVersionLabel, description string) (string, *versionutil.Version, error) {
+	// Do not print anything if description is ""
+	if description != "" {
+		fmt.Fprintf(g.w, "[upgrade/versions] Latest %s: ", description)
+	}
+
+	versionStr, err := kubeadmutil.KubernetesReleaseVersion(ciVersionLabel)
+	if err != nil {
+		if description != "" {
+			fmt.Fprintln(g.w, "<notfound>")
```

Comment:
yes. Since the previous print was intentionally a `Fprintf` without an `\n`, this is needed to inform about the error and make a new line in order for us to be able to print a normal line on the next print.